### PR TITLE
Add timezone to created_at

### PR DESCRIPTION
**Pull request by scosman**
We stored in local time, which isn't great.

 - add timezone so we know absolute time and user local time
 - find all compare cases (can't compare native to absolute)
 - tests

## Summary by CodeRabbit

* **Improvements**
  * Timestamps are now timezone-aware across the app, improving consistency and accuracy of displayed dates/times.
  * Cache staleness and timestamp comparisons use timezone-aware values to avoid cross-timezone inconsistencies.
  * Model validation now normalizes legacy timestamps so stored datetimes retain timezone information.

* **Documentation**
  * New specs and plans describing the timezone-aware datetime rollout.

* **Tests**
  * Expanded tests for parsing, formatting, normalization, and frontend date formatting.

### CODE REVIEW

File: libs/core/kiln_ai/datamodel/test_basemodel.py [hunk 1]
```diff
@@ -1371,3 +1371,60 @@ def test_all_children_of_parent_path_single_parent_type(tmp_path):
...
+    assert model.created_at.tzinfo is not None
+
+
+def test_created_at_roundtrip_preserves_offset(tmp_path):
+    model = KilnBaseModel(path=tmp_path / "test.kiln")
+    model.save_to_file()
+
+    raw_json = json.loads(model.path.read_text())
+    raw_ts = raw_json["created_at"]
+    assert "+" in raw_ts or "-" in raw_ts[10:]
```
Comment (leonardmq) [inline libs/core/kiln_ai/datamodel/test_basemodel.py]: This is the test failing on CI - it is flaky because when user is on UTC (as is the case with the CI), the date is serialized as `2026-04-17T05:17:14.731895Z` (no `+` in it)

I think this should fix it:
```suggestion
    assert raw_ts.endswith("Z") or "+" in raw_ts or "-" in raw_ts[10:]
```

**Review by leonardmq — APPROVED** [review]
Tested and it works - you have one test that fails on CI because UTC tz is a special case

File: libs/core/kiln_ai/datamodel/basemodel.py [hunk 2]
```diff
@@ -308,9 +309,22 @@ class KilnBaseModel(BaseModel):
...
+    @classmethod
+    def _normalize_created_at_tz(cls, v: Any) -> Any:
+        if isinstance(v, datetime) and v.tzinfo is None:
+            return v.astimezone()
+        if isinstance(v, str):
+            parsed = datetime.fromisoformat(v)
+            if parsed.tzinfo is None:
+                return parsed.astimezone()
+            return parsed
+        return v
```
[the remark this answers is not in the document]
Comment (scosman) [reply libs/core/kiln_ai/datamodel/basemodel.py]: Fixed in 45446693.

Switched the validator to `mode="after"` as suggested. The body now only ensures the resulting `datetime` is tz-aware, letting Pydantic's parser handle all valid ISO-8601 input formats (including `Z`-suffixed strings on Python 3.10) and numeric coercion.

---
*Addressed by AI coding agent via `/spec pr`*

File: libs/core/kiln_ai/datamodel/basemodel.py [hunk 3]
```diff
@@ -308,9 +309,22 @@ class KilnBaseModel(BaseModel):
...
+    @field_validator("created_at", mode="before")
+    @classmethod
+    def _normalize_created_at_tz(cls, v: Any) -> Any:
+        if isinstance(v, datetime) and v.tzinfo is None:
+            return v.astimezone()
+        if isinstance(v, str):
+            parsed = datetime.fromisoformat(v)
+            if parsed.tzinfo is None:
+                return parsed.astimezone()
+            return parsed
```
[the remark this answers is not in the document]
Comment (scosman) [reply libs/core/kiln_ai/datamodel/basemodel.py]: Fixed in 45446693.

Resolved by switching the validator to `mode="after"` (per the parallel Gemini suggestion). Pydantic's built-in parser handles the `Z` suffix on all supported Python versions, so we don't need the manual `Z → +00:00` normalization.

---
*Addressed by AI coding agent via `/spec pr`*

File: specs/projects/timezone_aware_datetimes/architecture.md [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+- `datetime.astimezone()` (no arg) on a naive datetime treats it as local and attaches the local TZ — exactly what we want for the legacy assumption.
+
+### Other datetime fields
+
+Audit Phase 1 deliverable identifies every other datetime field across the datamodel directory. For each, decide:
+
+- If it represents "moment something happened" → apply same default-factory + validator pattern. Hoist the validator into a reusable helper if more than one field needs it.
+- If it's an in-memory cache timestamp (e.g., `provider_api.py: AvailableModelsCache.last_updated`) → also fix the *comparison* call site (`datetime.now() - self.last_updated`) to use aware "now". Don't necessarily change the field type if it's never serialized.
+
+`provider_api.py` is **outside** the data model directory but the fix-pattern is identical and small — handle it in the consumer-fix phase.
```
[the remark this answers is not in the document]
Comment (scosman) [reply specs/projects/timezone_aware_datetimes/architecture.md]: Fixed in 45446693.

Updated the stale references: `AvailableModelsCache` → `OpenAICompatibleProviderCache`, and the comparison/initialization examples now use `datetime.now().astimezone()` to match the aware-now math in `provider_api.py`.

---
*Addressed by AI coding agent via `/spec pr`*

File: specs/projects/timezone_aware_datetimes/functional_spec.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+- Pydantic v2 default datetime serialization preserves the offset; no custom serializer required, but a round-trip test must lock this.
+
+### Reading an existing model instance from disk
+
+When a JSON file is loaded into a Pydantic model:
+
+- **Aware datetime in the JSON (has offset or `Z`):** parsed as-is. Awareness preserved.
+- **Naive datetime in the JSON (no offset, no `Z`):** the parse layer MUST attach the system's *local* timezone before any downstream code touches the value. This makes legacy single-user data render correctly (writer == reader).
+
+This must be enforced at the model boundary so all downstream consumers see only aware datetimes — no aware/naive comparison risk leaking out of the data layer.
```
[the remark this answers is not in the document]
Comment (scosman) [reply specs/projects/timezone_aware_datetimes/functional_spec.md:24]: Fixed in 45446693.

Reconciled the contract: the API now follows the model-boundary rule (normalize naive → aware at model parse), so responses always emit aware datetimes with an explicit offset. Updated the API output section to match.

---
*Addressed by AI coding agent via `/spec pr`*

File: specs/projects/timezone_aware_datetimes/functional_spec.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+- Legacy naive ISO → JS `Date` parses as viewer's local → renders correctly for single-user. ✓
+
+Behavior must be locked with tests covering all three formats.
+
+## Edge Cases
+
+- **User changed system TZ or traveled between writes (legacy naive data):** old naive timestamp is interpreted as *current* local TZ on read. May display ~hours off. Acknowledged, not fixed (no data to recover the original zone).
+- **Shared dataset with mixed-version collaborators (legacy naive data):** a naive timestamp written by User A in TZ X, read by User B in TZ Y, will display in B's local TZ as if it were B-local. Already broken today; this project does not fix legacy data.
+- **Future shared collaboration (new aware data):** the absolute instant is always correct. Both collaborators see correct relative times.
+- **Pydantic field assigned a naive datetime in code (e.g., test fixture):** must be caught. Either the parse normalizer also runs on assignment (preferred — uses `validate_assignment=True` already on `KilnBaseModel`), or tests are updated to write aware datetimes. Architecture decides.
```
[the remark this answers is not in the document]
Comment (scosman) [reply specs/projects/timezone_aware_datetimes/functional_spec.md]: Fixed in 45446693.

Replaced "Architecture decides" with the concrete requirement: `KilnBaseModel` sets `validate_assignment=True`, so assigning a naive datetime runs the same normalizer and is either coerced to aware or raises.

---
*Addressed by AI coding agent via `/spec pr`*

File: libs/core/kiln_ai/datamodel/test_basemodel.py (hunk 1 above)
Comment (scosman) [reply libs/core/kiln_ai/datamodel/test_basemodel.py]: Fixed in 45446693.

Applied your suggested assertion. Good catch on the UTC-runner case — Pydantic serializes `+00:00` as `Z` on the CI host, so the offset check needed to accept `endswith("Z")` too.

---
*Addressed by AI coding agent via `/spec pr`*
